e2e/tests/upgrades/v6/upgrade_test.go

```diff
@@ -0,0 +1,184 @@
+package v6
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"testing"
+	"time"
+
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	"github.com/strangelove-ventures/ibctest/v6/chain/cosmos"
+	"github.com/strangelove-ventures/ibctest/v6/ibc"
+	"github.com/strangelove-ventures/ibctest/v6/test"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/cosmos/ibc-go/e2e/testconfig"
+	"github.com/cosmos/ibc-go/e2e/testsuite"
+	"github.com/cosmos/ibc-go/e2e/testvalues"
+)
+
+const (
+	haltHeight         = uint64(100)
+	blocksAfterUpgrade = uint64(10)
+)
+
+func TestUpgradeV6TestSuite(t *testing.T) {
+	suite.Run(t, new(UpgradeV6TestSuite))
+}
+
+type UpgradeV6TestSuite struct {
+	testsuite.E2ETestSuite
+}
+
+// UpgradeChain upgrades a chain to a specific version using the planName provided.
+// The software upgrade proposal is broadcast by the provided wallet.
+func (s *UpgradeV6TestSuite) UpgradeChain(ctx context.Context, chain *cosmos.CosmosChain, wallet *ibc.Wallet, planName, upgradeVersion string) {
+	prevVersion := chain.Nodes()[0].Image.Version
+	plan := upgradetypes.Plan{
+		Name:   planName,
+		Height: int64(haltHeight),
+		Info:   fmt.Sprintf("upgrade version test from %s to %s", prevVersion, upgradeVersion),
+	}
+	upgradeProposal := upgradetypes.NewSoftwareUpgradeProposal(fmt.Sprintf("upgrade from %s to %s", prevVersion, upgradeVersion), "upgrade chain E2E test", plan)
+	s.ExecuteGovProposal(ctx, chain, wallet, upgradeProposal)
+
+	height, err := chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height before upgrade")
+
+	timeoutCtx, timeoutCtxCancel := context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
+
+	err = test.WaitForBlocks(timeoutCtx, int(haltHeight-height)+1, chain)
+	s.Require().Error(err, "chain did not halt at halt height")
+
+	err = chain.StopAllNodes(ctx)
+	s.Require().NoError(err, "error stopping node(s)")
+
+	chain.UpgradeVersion(ctx, s.DockerClient, upgradeVersion)
+
+	err = chain.StartAllNodes(ctx)
+	s.Require().NoError(err, "error starting upgraded node(s)")
+
+	timeoutCtx, timeoutCtxCancel = context.WithTimeout(ctx, time.Minute*2)
+	defer timeoutCtxCancel()
+
+	err = test.WaitForBlocks(timeoutCtx, int(blocksAfterUpgrade), chain)
+	s.Require().NoError(err, "chain did not produce blocks after upgrade")
+
+	height, err = chain.Height(ctx)
+	s.Require().NoError(err, "error fetching height after upgrade")
+
+	s.Require().GreaterOrEqual(height, haltHeight+blocksAfterUpgrade, "height did not increment enough after upgrade")
+}
+
+// TODO: Swap e2e transfer flow logic for interchain accounts
```

Comment:
Refactor the transfer logic to follow ICS27 workflow.

- [ ] Create an ICS27 channel using an existing auth module. E.g. intertx
- [ ] Verify SendTx behaves as expected
- [x] Create software upgrade proposal
- [x] Execute proposal and upgrade to new chain binary
- [x] Migrations are run via upgrade handler and automatic migrations via consensus version increment.
- [ ] Verify SendTx behaves as expected